"""
This script was written by Kelsey Witt in March 2018. This file takes an angsd .beagle genotype
likelihood file (generated using the command angsd angsd -GL 1 -doGlf 2) and filters it to only 
include one SNP per window using the unlinked_snps.txt file (generated using angsd_snp_windows_selection.py).
The output is another .beagle file, just with fewer SNPs.

Usage: python3 filter_angsd_beagle_snps.py

The variable "snp_infile" must be changed to the name of the file generated by angsd_snp_windows_selection.py,
and the variable "beagle_infile" must be changed to the name of your .beagle file. The variable "outfile"
can also be changed.

Contact: kelsey_witt_dillon@brown.edu
"""

snp_infile = "macaque_gl_1p_50_01_unlinked_snps.txt"
beagle_infile = "macaque_gl_1p_50_01.beagle"
outfile = "macaque_gl_1p_50_01_unlinked_snps.beagle"

SNPs_to_Keep = {}
with open(snp_infile) as f:
    for line in f:
        line_split = line.split()
        chromosome, position = line_split[1:3]
        if chromosome not in SNPs_to_Keep.keys():
            SNPs_to_Keep[chromosome]=[]
        SNPs_to_Keep[chromosome].append(position)

with open(outfile, 'w') as g:
    with open(beagle_infile) as h:
        for line in h:
            if "marker" in line:
                g.write(line)
            else:
                line_split = line.split()
                chrom_pos = line_split[0].split("_")
                b_chr, b_pos = chrom_pos[0:2]
                if b_pos in SNPs_to_Keep[b_chr]:
                    g.write(line)

